### PR TITLE
[InstCombine] Don't consider aligned_alloc removable if icmp uses result

### PR DESCRIPTION
**Pull request by fhahn**
At the moment, all alloc-like functions are assumed to return non-null pointers, if their return value is only used in a compare. This is based on being allowed to substitute the allocation function with one that doesn't fail to allocate the required memory.

aligned_alloc however must also return null if the required alignment cannot be satisfied, so I don't think the same reasoning as above can be applied to it.

This patch adds a bail-out for aligned_alloc calls to isAllocSiteRemovable.

### CODE REVIEW

**Comment by nikic** [comment]
I don't think you can bail out for all aligned allocations here. For example, Rust always uses aligned allocations (via `__rust_alloc`) and will always pass a valid alignment, so I don't think we need to consider that failure mode there.

This should probably be hardcoded to `aligned_alloc` in particular.

**Comment by fhahn** [comment]
Updated the patch to only bail out for `aligned_alloc` and not bail out if alignment and size are known good (alignment is a power of 2 and size is a multiple of alignment)

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 1]
```diff
@@ -2430,6 +2430,30 @@ static bool isAllocSiteRemovable(Instruction *AI,
         unsigned OtherIndex = (ICI->getOperand(0) == PI) ? 1 : 0;
         if (!isNeverEqualToUnescapedAlloc(ICI->getOperand(OtherIndex), TLI, AI))
           return false;
+        // Do not fold compares to aligned_alloc calls, as they may have to
+        // return null in case the required alignment cannot be satisfied,
+        // unless we can prove that both alignment and size are valid.
+        if ((getAllocFnKind(AI) & AllocFnKind::Aligned) !=
+            AllocFnKind::Unknown) {
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2437]: Do we still need this check if we're also checking TLI? (Move the TLI check up here?)

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 2]
```diff
@@ -2430,6 +2430,30 @@ static bool isAllocSiteRemovable(Instruction *AI,
         unsigned OtherIndex = (ICI->getOperand(0) == PI) ? 1 : 0;
         if (!isNeverEqualToUnescapedAlloc(ICI->getOperand(OtherIndex), TLI, AI))
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2431]: I was going to say that the new checks should be in the  `isa<ConstantPointerNull>` path of isNeverEqualToUnescapedAlloc(), but on further consideration, they also apply to the other two cases (e.g. you could compare the result of two required-to-be-null aligned_allocs), so current code structure is good.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp]: Not really, removed them

File: llvm/include/llvm/Analysis/MemoryBuiltins.h [hunk 3]
```diff
@@ -131,6 +131,11 @@ Constant *getInitialValueOfAllocation(const Value *V,
 std::optional<StringRef> getAllocationFamily(const Value *I,
                                              const TargetLibraryInfo *TLI);
 
+/// If \p V is a call of a function part of an allocation family e.g.
+/// malloc/realloc/calloc/free), return the allocation function kind the called
+/// function.
+AllocFnKind getAllocFnKind(const Value *V);
```
Comment (nikic) [inline llvm/include/llvm/Analysis/MemoryBuiltins.h:137]: These changes are no longer necessary?

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/include/llvm/Analysis/MemoryBuiltins.h (hunk 3 above)
Comment (fhahn) [reply llvm/include/llvm/Analysis/MemoryBuiltins.h]: Removed, thanks!

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 4]
```diff
@@ -2430,6 +2430,26 @@ static bool isAllocSiteRemovable(Instruction *AI,
...
+          // Check if alignment and size of a call to aligned_alloc is valid,
+          // that is alignment is a power-of-2 and the size is a multiple of the
+          // alignment.
+          const APInt *Alignment;
+          const APInt *Size;
+          return match(CB->getArgOperand(0), m_APInt(Alignment)) &&
+                 match(CB->getArgOperand(1), m_APInt(Size)) &&
+                 Alignment->isPowerOf2() && Size->urem(*Alignment).isZero();
+        };
+        auto *CB = dyn_cast<CallBase>(AI);
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2447]: Is this CB different from one the passed to the lambda?

**Pull request merged**